### ISSUE TITLE
Variable rate phase vocoding

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#517]
`phase_vocoder` currently accepts a single scalar parameter controlling the rate of time dilation.  In principle, it shouldn't be difficult to either vectorize this, or support a slice+rate parameterization so that different time regions can be accelerated by different rates.

**Comment by dpwe** [comment]
Once you get to variable rates, you have to make sure you are clear about
how the time base is defined, because now the input and output time bases
have a nonlinear relationship.  What you really want is the mapping from
output timebase to input timebase, since it's not infeasible to want
reversals and loops (non-monotonic relationship).  If you accept a vector
of per-input-window time scale, then effectively you're specifying how much
time each fixed-size input window will occupy in the output, and it's hard
to predict how long the output will be.

I think accepting a vector of (input time, output time) (defining a
piecewise-linear relationship) tuples is a nice, neutral description.  You
could then programmatically figure out the slope (time scaling) and time
center required from the input signal for each output frame.

**Comment by dpwe** [comment]
See for example:

https://github.com/dpwe/pitchfilter/blob/master/resample_map.m

(which, weirdly, has been translated into Matlab somewhere between my
memory and GitHub).  That code does "time scaling" by simple resampling,
but otherwise it's a similar concept.

  DAn.

**Comment by bmcfee** [comment]
> I think accepting a vector of (input time, output time) (defining a
> piecewise-linear relationship) tuples is a nice, neutral description.

Yeah, I was just thinking in terms of slices because then the bounds are explicit rather than implicit, and rate because it's closest to the existing interface.

This issue was brought up at the HAMR in rochester, where someone wanted an easy way to warp audio along a dtw path.  I had implemented something along these lines way back as part of the frankenmasher, but never migrated the code into librosa because it was kludgy as all get out.

**Comment by stefan-balke** [comment]
That would be great!

Referencing this: https://github.com/librosa/librosa/issues/298#issuecomment-201908080
and the implementation in the TSM Toolbox: https://www.audiolabs-erlangen.de/content/resources/MIR/TSMtoolbox/pvTSM.m

If we could the following points, we would pretty much cover the capabilities of the TSM Toolbox:
* HPSS ✅ 
* H-Part: Phase Vocoder ✅ 
* P-Part: Overlap-Add ❎ 
* Add up everything ✅ 

That could mean to have some `tsm(x, factor=None, wp=[tuples], ...)` frontend function.

**Comment by bmcfee** [comment]
pyrubberband now supports variable-rate stretching, and generally sounds pretty good.  I still think it's worth adding overlap-add stretching for illustrative purposes, but there's not much need at this point for a pure python variable-rate stretcher.

Any objection to closing this out?

**Comment by stefan-balke** [comment]
Hmm...would still like to see it in `librosa`. Give me time until the end of the year. :)

The thing with `rubberband` is that code is well-written but (at least for me), it takes some time to find the relevant parts...

**Comment by bmcfee** [comment]
> Give me time until the end of the year. :)

:bellhop_bell:
